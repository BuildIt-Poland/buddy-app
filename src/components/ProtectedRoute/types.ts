import React from 'react';
import { RouteProps } from 'react-router';

export type ProtectedRouteProps = { component: React.ComponentType<any> } & Partial<
  RouteProps
>;
