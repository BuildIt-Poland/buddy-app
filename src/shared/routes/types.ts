import React from 'react';

export type RouteProps = {
  url: string;
  component: React.ComponentType<any>;
};

export type Routes = {
  [key: string]: RouteProps | { [key: string]: RouteProps };
};
