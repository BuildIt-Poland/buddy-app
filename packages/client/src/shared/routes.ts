enum ROUTES {
  BASE = '/',
  LOGIN = '/login',
  FORGOT_PASSWORD = '/forgot-password',
  RESET_PASSWORD = '/reset-password/:token',
  ROUTE_404 = '/404',

  TALENT = '/talent',
  TALENT_SELECT_BUDDY = '/talent/buddies',
  TALENT_SELECT_NEWBIE = '/talent/buddies/:buddyId/newbies',
  TALENT_ADD_TASK = '/talent/newbies/:newbieId/add-task',
  TALENT_TASK_DETAILS = '/talent/newbies/:newbieId/tasks/:taskId',
  TALENT_TASKS_LIST = '/talent/newbies/:newbieId/tasks',
  TALENT_ADD_NEWBIE = '/talent/buddies/:buddyId/add-newbie',
  TALENT_ADD_BUDDY = '/talent/add-buddy',
  TALENT_ADD_TALENT = '/talent/add-talent',
  TALENT_DETAILS = '/talent/details',
  TALENT_BUDDY_DETAILS = '/talent/buddies/:buddyId/details',
  TALENT_NEWBIE_DETAILS = '/talent/newbies/:newbieId/details',

  BUDDY = '/buddy/',
  BUDDY_DETAILS = '/buddy/details',
  BUDDY_ADD_TASK = '/buddy/newbies/:newbieId/add-task',
  BUDDY_TASK_DETAILS = '/buddy/newbies/:newbieId/tasks/:taskId',
  BUDDY_TASKS_LIST = '/buddy/newbies/:newbieId/tasks',
  BUDDY_SELECT_NEWBIE = '/buddy/newbies',
  BUDDY_ADD_NEWBIE = '/buddy/add-newbie',
  BUDDY_NEWBIE_DETAILS = '/buddy/newbies/:newbieId/details',

  NEWBIE = '/newbie',
  NEWBIE_DETAILS = '/newbie/details',
  NEWBIE_TASKS_LIST = '/newbie/tasks',
  NEWBIE_TASK_DETAILS = '/newbie/tasks/:taskId',
  NEWBIE_BUDDY_DETAILS = '/newbie/buddy/:buddyId/details',
}

export { ROUTES };
