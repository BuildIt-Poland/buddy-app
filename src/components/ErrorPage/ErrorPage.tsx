import React from 'react';
import Typography from '@material-ui/core/Typography';
import { ReactComponent as R2d2 } from 'assets/svg/r2d2.svg';
import Button from '@material-ui/core/Button';
import Grid from '@material-ui/core/Grid';
import makeStyles from '@material-ui/core/styles/makeStyles';
import BackGroundShape from 'components/BackgroundShape';
import Container from '@material-ui/core/Container';
import { ErrorPageProps } from './types';
import ERROR_PAGE_DICTIONARY from './dictionary';

const useStyles = makeStyles(theme => ({
  root: {
    height: '100%',
    display: 'flex',
  },
  r2d2: {
    animation: `$float 3s ${theme.transitions.easing.easeInOut} infinite`,
  },
  '@keyframes float': {
    '0%': {
      transform: 'translateY(1rem)',
    },
    '50%': {
      transform: 'translateY(-1rem)',
    },
    '100%': {
      transform: 'translateY(1rem)',
    },
  },
}));

const ErrorPage: React.FC<ErrorPageProps> = ({ title, message, ...props }) => {
  const classes = useStyles();

  return (
    <Container
      component={'main'}
      className={classes.root}
      maxWidth={'md'}
      {...props}>
      <Grid
        spacing={1}
        container
        direction='column'
        justify='center'
        alignItems='center'>
        <Grid item>
          <Typography component='h1' variant='h1' align={'center'}>
            {title}
          </Typography>
        </Grid>
        <Grid item>
          <Typography component={'p'} align={'center'}>
            {message}
          </Typography>
        </Grid>
        <Grid item>
          <R2d2 className={classes.r2d2} />
        </Grid>
        <Grid item>
          <Button href={'/'} variant={'outlined'} color='primary'>
            {ERROR_PAGE_DICTIONARY.GO_HOMEPAGE}
          </Button>
        </Grid>
        <BackGroundShape />
      </Grid>
    </Container>
  );
};

export default ErrorPage;
