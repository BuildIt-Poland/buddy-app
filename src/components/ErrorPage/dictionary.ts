interface ErrorPageDictionary {
  GO_HOMEPAGE: string;
}

const ERROR_PAGE_DICTIONARY: ErrorPageDictionary = {
  GO_HOMEPAGE: 'Go to the homepage',
};

export default ERROR_PAGE_DICTIONARY;
