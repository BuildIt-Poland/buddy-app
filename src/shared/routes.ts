enum ROUTES {
  BASE = '/',
  LOGIN = '/login',
  ROUTE_404 = '/404',
  BUDDY_ADD_TASK = '/buddy/newbies/:newbieId/add-task',
  BUDDY_TASK_DETAILS = '/buddy/newbies/:newbieId/tasks/:taskId',
  BUDDY_TASKS_LIST = '/buddy/newbies/:newbieId/tasks',
  BUDDY_SELECT_NEWBIE = '/buddy/newbies',
  BUDDY_ADD_NEWBIE = '/buddy/add-newbie',
  BUDDY_NEWBIE_DETAILS = '/buddy/newbies/:newbieId/details',
  NEWBIE_TASK_DETAILS = '/newbie/tasks/:taskId',
  NEWBIE_TASKS_LIST = '/newbie/tasks',
  NEWBIE_BUDDY_DETAILS = '/newbie/buddy/:buddyId/details',
}

export { ROUTES };
