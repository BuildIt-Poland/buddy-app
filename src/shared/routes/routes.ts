enum ROUTES {
  BASE = '/',
  LOGIN = '/login',
  ERROR = '/404',
  BUDDY_ADD_TASK = '/buddy/add-task',
  BUDDY_TASK_DETAILS = '/buddy/tasks/:taskId',
  BUDDY_TASKS_LIST = '/buddy/tasks',
  BUDDY_SELECT_NEWBIE = '/buddy/select-newbie',
  BUDDY_ADD_NEWBIE = '/buddy/add-newbie',
  BUDDY_NEWBIE_DETAILS = '/buddy/newbie/:id/details',
  NEWBIE_TASK_DETAILS = '/newbie/tasks/:taskId',
  NEWBIE_TASKS_LIST = '/newbie/tasks',
}

export { ROUTES };
