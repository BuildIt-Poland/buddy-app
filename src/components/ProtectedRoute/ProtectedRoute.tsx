import React from 'react';
import { Route, Redirect } from 'react-router-dom';
import userService from '../../utils/user.service';
import { ROUTES } from '../../shared/routes';

const ProtectedRoute: React.FC<any> = ({ component: Component, ...rest }) => {
  return (
    <Route
      {...rest}
      render={props =>
        userService.isAuthenticated ? (
          <Component {...props} />
        ) : (
          <Redirect
            to={{
              pathname: ROUTES.LOGIN,
            }}
          />
        )
      }
    />
  );
};

export default ProtectedRoute;
