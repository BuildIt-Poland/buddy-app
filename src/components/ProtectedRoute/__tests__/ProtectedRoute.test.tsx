import React from 'react';
import { create } from 'react-test-renderer';
import { MemoryRouter } from 'react-router';
import ProtectedRoute from '../ProtectedRoute';
import userService from '../../../utils/user-service';
import { ROUTES } from '../../../shared/routes';
import ErrorPage from '../../ErrorPage';

jest.mock('../../../utils/user-service', () => ({
  isAuthenticated: jest.fn(),
  getUser: jest.fn(),
}));

describe('Component - ProtectedRoute', () => {
  describe('when user is authorised', () => {
    beforeAll(() => {
      (userService.isAuthenticated as jest.Mock).mockReturnValue(true);
    });
    test('route component renders correctly', () => {
      const component = create(
        <MemoryRouter initialEntries={[ROUTES.ERROR]}>
          <ProtectedRoute path={ROUTES.ERROR} exact component={ErrorPage} />
        </MemoryRouter>
      );
      expect(component.toJSON()).toMatchSnapshot();
    });
  });
  describe('when user is unauthorised', () => {
    beforeAll(() => {
      (userService.isAuthenticated as jest.Mock).mockReturnValue(false);
    });
    test('route component is not rendered', () => {
      const component = create(
        <MemoryRouter initialEntries={[ROUTES.ERROR]}>
          <ProtectedRoute path={ROUTES.ERROR} exact component={ErrorPage} />
        </MemoryRouter>
      );
      expect(component.toJSON()).toMatchSnapshot();
    });
  });
});
