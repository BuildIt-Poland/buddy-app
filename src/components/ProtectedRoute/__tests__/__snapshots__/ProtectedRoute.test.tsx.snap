// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - ProtectedRoute when user is authorised route component renders correctly 1`] = `
<h1>
  Error
</h1>
`;

exports[`Component - ProtectedRoute when user is unauthorised route component is not rendered 1`] = `null`;
